### PR TITLE
tapfreighter: resume pending parcels using the main porter goroutine

### PR DESCRIPTION
**Pull request by ffranr**
In this commit, we ensure that pending parcels are resumed using the same goroutine that processes the normal send parcel processing flow.

The goal of this change is to ensure that the locking mechanism enforced by a singular parcel processing goroutine is not compromised by the execution of a secondary parcel goroutine.

### CODE REVIEW

File: tapfreighter/chain_porter.go [hunk 1]
```diff
@@ -116,31 +116,39 @@ func (p *ChainPorter) Start() error {
...
-		// Now that we have the set of pending sends, we'll make a new
-		// goroutine that'll drive the state machine till the broadcast
-		// point (which we might be repeating), and final terminal
-		// state.
+		// At this point the asset porter should be running, so it will
+		// pick up the pending parcels and attempt to deliver them.
 		for _, parcel := range pendingParcels {
-			p.Wg.Add(1)
-			go p.resumePendingParcel(parcel)
+			p.exportReqs <- parcel
```
Comment (guggero) [inline tapfreighter/chain_porter.go:150]: Why not just:
```
		for idx := range outboundParcels {
			p.exportReqs <- NewPendingParcel(outboundParcels[idx])
		}
```

Then we don't need the loop above. And fix a potential loop variable issue.

Comment (guggero) [reply tapfreighter/chain_porter.go:150]: Also, should we add the log message back here?
```
	log.Infof("Attempting to resume delivery for anchor_txid=%v",
		pkg.AnchorTx.TxHash().String())
```

**Review by guggero — COMMENTED** [review]
Very nice fix!

Does that mean you were successful in reproducing the error described in https://github.com/lightninglabs/taproot-assets/issues/332? If yes, should we link the issue to the PR so it's closed by it?

**Comment by jharveyb** [comment]
Started testing this manually with scenario 2 from here: https://github.com/lightninglabs/taproot-assets/issues/332#issuecomment-1588932991

I think we can actually fix this another way. If we ignore the error `output already spent` from the `PublsihTransaction` call in `SendStateBroadcast` here:

https://github.com/lightninglabs/taproot-assets/blob/f46f3f10d9032e87a8cab2d2a328e8aa6334b242/tapgarden/caretaker.go#L716

Then, for already confirmed transactions, we should recognize the confirmation in the next state and continue as normal instead of shutting down the caretaker. As-is I think we would still fail if the send TX already confirmed.

**Comment by guggero** [comment]
> If we ignore the error output already spent from the PublsihTransaction call in SendStateBroadcast here

I don't think we want to ignore the error in the current configuration. The `output already spent` error happens because we don't have any asset coin locking implemented yet. So we avoid running into problems by only having a single goroutine active for issuing transfers. So a user is forced to wait between sends until the first one confirms.
If a user instead restarts the daemon before the first transfer is confirmed, the restart continues the transfer in a separate goroutine, which makes it possible for another one being issued through the RPC in the meantime. And because we don't have asset coin locking, it will select the same on-chain input for the second transfer and it will never go through.
This PR fixes that issue by using the same goroutine for resuming parcels that is also used for handling new parcels from RPC requests.

**Comment by jharveyb** [comment]
That makes sense; but I think with these changes the resume of the parcel will still fail if the send TX was already confirmed, and we won't reach the later states of the send logic?

I think maybe we would want to check for send TX confirmation and jump past re-broadcasting if confirmed.

I'll read more closely + try and repro again.

**Comment by ffranr** [comment]
> we would want to check for send TX confirmation and jump past re-broadcasting if confirmed

That sounds right to me @jharveyb  !

**Comment by jharveyb** [comment]
> > we would want to check for send TX confirmation and jump past re-broadcasting if confirmed
> 
> That sounds right to me @jharveyb !

We already have the appropriate logic in `waitForTransferTxConf`, but I'm not sure if that will pick up a confirmation in a previous block; maybe we can request a scan forward from the height hint to the present block (do we care if the TX is not already confirmed, but already in the mempool? I think not), and then move to broadcast if that fails instead of waiting for a confirmation that would not arrive.

**Comment by ffranr** [comment]
@jharveyb @guggero I think we should still merge this PR regardless. I think it simplifies the handling of pending packages. I don't see why we should start a goroutine for each pending package, feels like premature optimisation.

**Review by guggero — APPROVED** [review]

**Review by jharveyb — APPROVED** [review]

**Pull request merged**